### PR TITLE
【开源实习】基于MindSpore NLP实现language modeling应用案例开发

### PR DESCRIPTION
**Pull request by 4everWZ**
# 基于 MindSpore MindNLP 的语言模型微调实战（DistilGPT-2 Causal LM）

本 Notebook 演示如何在 **MindSpore 2.7.0** 和 **MindNLP (Mindhf) 0.5.1** 环境下，参考 HuggingFace 官方 `language_modeling` 流程，对 **DistilGPT-2** 模型进行微调，并完成文本生成推理。

与官方示例不同的是，本项目针对 MindSpore 静态图/动态图特性，对 `Trainer` 进行了轻量级适配，使其既能享受 HF 风格的 API 便利性，又能稳定运行于 Ascend/GPU 环境。

## 主要内容

1. **数据集加载**：
   - 使用 MindNLP 的 `load_dataset` 接口从 HuggingFace Datasets 镜像源加载 **Wikitext-2-raw-v1** 数据集。
2. **模型与分词器**：
   - 使用 `AutoTokenizer` 和 `AutoModelForCausalLM` 加载轻量级的 **DistilGPT-2** 预训练模型。
   - 处理 GPT-2 家族特有的 padding 问题（将 `pad_token` 设为 `eos_token`）。
3. **数据处理与适配**：
   - 构造 Causal LM 训练样本（`input_ids` 与 `labels` 同步偏移）。
   - **关键适配**：实现 `MSMapDataset` 类，将 MindSpore 的流式 Dataset 封装为 Map-style 格式，并配合 `passthrough_collator`，解决 Trainer 内部 DataLoader 的兼容性问题。
4. **模型微调（自定义 Trainer）**：
   - 未使用底层的原生循环，而是继承并实现 **`NoJitTrainer`**。
   - **核心逻辑**：重写 `training_step`，移除默认的 JIT 编译依赖，采用显式 `.backward()` 方式更新梯度。这解决了 `value_and_grad` 在部分复杂控制流下的兼容性报错，同时保留了 `TrainingArguments` 的便捷配置（如 `logging_steps`、`gradient_accumulation` 等）。
5. **推理与生成**：
   - 使用 `model.generate` 接口进行 Temperature 采样与 Top-P 解码，实现高质量的文本自动续写。

### CODE REVIEW

**Comment by 4everWZ** [comment]
已经在Ascend硬件中finetune过了，已经更新pr了
等老师review后，会去squash

**Comment by moyu026** [comment]
使用mindspore2.7.0和mindnlp0.5.1可以运行，但是需要修改diffusers版本，pip install diffusers==0.35.2

**Comment by 4everWZ** [comment]
> 使用mindspore2.7.0和mindnlp0.5.1可以运行，但是需要修改diffusers版本，pip install diffusers==0.35.2

是的，diffuser的版本在mindnlp有一个issue有提到过这个问题。
https://github.com/mindspore-lab/mindnlp/issues/2306

**Comment by moyu026** [comment]
请参考这个模板再修改下格式，https://github.com/BQBBLZ/applications/blob/master/ai_x/train_resnet_classification.ipynb

**Comment by 4everWZ** [comment]
> 请参考这个模板再修改下格式，https://github.com/BQBBLZ/applications/blob/master/ai_x/train_resnet_classification.ipynb

您好，已经修改格式了，请您检查一下

**Comment by xing-yiren** [comment]
1. 文件中涉及MindNLP名称建议改成全称MindSpore NLP
2. 这个案例放到llm目录下吧，把文件名改成模型名称（模型名称可以参考huggingface transformers库）的命名方法，同步更新下目录，辛苦了
3. ipynb文件名参考[wiki](https://github.com/mindspore-lab/applications/wiki/Contributing-Guidelines#43-notebook-naming-convention)的规范修改下

**Comment by 4everWZ** [comment]
> notebook的中间运行结果产物也需要删除，原因如下：
> 
> 考虑到不同人运行 Notebook 会产生不同的execution_count、输出结果，提交后容易导致 Git 冲突；且Notebook 的输出（如图片、大段文本）会让文件体积变大，建议提交notebook前，通过nbstripout剔除运行时产生的动态数据，仅保留代码、Markdown 文本结构，让文件更干净
> 
> 具体操作： fork并clone代码仓后，执行： pip install nbstripout nbstripout --install 上述操作后，后续每一次提交时，Git 都会自动调用nbstripout清理.ipynb文件的过程数据
> 
> 也可以去此链接中查看全部的操作过程：https://github.com/mindspore-lab/applications/wiki/Contributing-Guidelines

已经删除输出结果了

**Comment by 4everWZ** [comment]
> 1. 文件中涉及MindNLP名称建议改成全称MindSpore NLP
> 2. 这个案例放到llm目录下吧，把文件名改成模型名称（模型名称可以参考huggingface transformers库）的命名方法，同步更新下目录，辛苦了
> 3. ipynb文件名参考[wiki](https://github.com/mindspore-lab/applications/wiki/Contributing-Guidelines#43-notebook-naming-convention)的规范修改下

已经将markdown中不影响导入安装的部分改成MindSpore NLP了，以及文件位置以及移动到llm/distilgpt2/finetune_distilgpt2_language_modeling.ipynb下了
已经rebase head了 只有一次commit记录

File: llm/distilgpt2/finetune_distilgpt2_language_modeling.ipynb [hunk 1]
```diff
@@ -0,0 +1,758 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "fe1b41a2",
+   "metadata": {},
+   "source": [
+    "#### **设置 MindSpore 上下文**\n",
+    "\n",
+    "Ascend（Atlas）设备推荐使用 GRAPH_MODE"
```
Comment (xing-yiren) [inline llm/distilgpt2/finetune_distilgpt2_language_modeling.ipynb]: 不好意思这里才注意到，应用案例推荐使用动态图来运行开发，辛苦做下修改吧

File: llm/distilgpt2/finetune_distilgpt2_language_modeling.ipynb [hunk 2]
```diff
@@ -0,0 +1,758 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "6ccef786",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# ===== 6. 使用 MindNLP Trainer 训练 =====\n",
```
Comment (xing-yiren) [inline llm/distilgpt2/finetune_distilgpt2_language_modeling.ipynb]: MindSpore NLP

**Comment by xing-yiren** [comment]
不过还是要重点夸夸，能用AI辅助代码检查，同时还能rebase head保证只有一次commit，优秀！

File: llm/distilgpt2/finetune_distilgpt2_language_modeling.ipynb (hunk 1 above)
Comment (4everWZ) [reply llm/distilgpt2/finetune_distilgpt2_language_modeling.ipynb]: 已经改成PYNATIVE_MODE了

File: llm/distilgpt2/finetune_distilgpt2_language_modeling.ipynb [hunk 3]
```diff
@@ -0,0 +1,781 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "7",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "device_target = os.getenv(\"DEVICE_TARGET\", \"Ascend\")  # 如需在CPU上运行，将\"Ascend\"改为\"CPU\"\n",
+    "print(\"Using device:\", device_target)\n",
+    "\n",
+    "context.set_context(\n",
```
Comment (xing-yiren) [inline llm/distilgpt2/finetune_distilgpt2_language_modeling.ipynb:136]: 现在默认运行走的就是动态图了，所以理论上这里不需要再显式配置了

File: llm/distilgpt2/finetune_distilgpt2_language_modeling.ipynb [hunk 4]
```diff
@@ -0,0 +1,781 @@
...
+   "cell_type": "markdown",
+   "id": "16",
+   "metadata": {},
+   "source": [
+    "## 模型训练\n",
+    "\n",
+    "我们采用 MindSpore 官方推荐的原生训练形式：\n",
+    "\n",
+    "1. 使用 `AdamW` 作为优化器；  \n",
+    "2. 使用 `value_and_grad` 构造带梯度的前向函数；\n",
```
Comment (xing-yiren) [inline llm/distilgpt2/finetune_distilgpt2_language_modeling.ipynb]: 我看代码里是手写了个trainer，没有用到`value_and_grad`这个接口，是不是文字这里需要同步改下

**Comment by xing-yiren** [comment]
README辛苦同步更新下案例信息，其他的没有了，辛苦辛苦

**Comment by xing-yiren** [comment]
@moyu026 登金辛苦也同步验收下动态图下运行是否没问题

**Comment by moyu026** [comment]
动态图可以运行

File: llm/distilgpt2/finetune_distilgpt2_language_modeling.ipynb (hunk 4 above)
Comment (4everWZ) [reply llm/distilgpt2/finetune_distilgpt2_language_modeling.ipynb]: 已经修改

File: llm/distilgpt2/finetune_distilgpt2_language_modeling.ipynb (hunk 3 above)
Comment (4everWZ) [reply llm/distilgpt2/finetune_distilgpt2_language_modeling.ipynb:136]: 已经删除

**Comment by xing-yiren** [comment]
目前我这边看没什么问题了，辛苦 @wang-hua-2019 把把关

**Review by xing-yiren — APPROVED** [review]

**Review by lvyufeng — APPROVED** [review]
